morl_baselines/multi_policy/capql/capql.py

```diff
@@ -0,0 +1,471 @@
+"""CAPQL algorithm."""
+import os
+import random
+from itertools import chain
+from typing import List, Optional, Union
+
+import gymnasium
+import numpy as np
+import torch as th
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+import wandb
+from torch.distributions import Normal
+
+from morl_baselines.common.evaluation import policy_evaluation_mo
+from morl_baselines.common.morl_algorithm import MOAgent, MOPolicy
+from morl_baselines.common.networks import mlp
+from morl_baselines.common.utils import (
+    equally_spaced_weights,
+    layer_init,
+    log_all_multi_policy_metrics,
+    log_episode_info,
+    polyak_update,
+)
+
+
+LOG_SIG_MAX = 2
+LOG_SIG_MIN = -20
+EPSILON = 1e-6
+
+
+class ReplayMemory:
+    """Replay memory."""
```

Comment:
Can't we use the implementation of replay buffer we have in the repo?

Reply:
They also push reward weights to the buffer. We could adapt our replay buffer in the future, but for now I think it is easier to just use the buffer in their original code.